src/pudl/output/pudltabl.py

```diff
@@ -1207,6 +1207,62 @@ def epacamd_eia(
             self._dfs["epacamd_eia"] = pudl.output.epacems.epacamd_eia(self.pudl_engine)
         return self._dfs["epacamd_eia"]
 
+    ###########################################################################
+    # FOR PICKLING AND OTHER IO
+    ###########################################################################
+
+    def __getstate__(self) -> dict:
+        """Get current object state for pickle.
+
+        This method is run as part of pickling the object. It needs to return the
+        object's current state with any un-serializable objects converted to a form that
+        can be serialized. See :meth:`object.__getstate__` for further details on the
+        expected behavior of this method.
+        """
+        return self.__dict__.copy() | {
+            # defaultdict may be serializable but lambdas are not, so it must go
+            "_dfs": dict(self.__dict__["_dfs"]),
+            # sqlalchemy engines are also a problem here, saving the URL should
+            # provide enough of what is needed to recreate it, though that means the
+            # pickle is not portable, but any fix to that should probably happen
+            # on the other side
+            "pudl_engine": str(self.__dict__["pudl_engine"].url)
+            .removeprefix("Engine(")
+            .removesuffix(")"),
+        }
+
+    def __setstate__(self, state: dict) -> None:
+        """Restore the object's state from a dictionary.
+
+        This method is run when the object is restored from a pickle. Anything
+        that was changed in :meth:`pudl.output.pudltabl.PudlTabl.__getstate__` must be
+        undone here. Another important detail is that ``__init__`` is not run when an
+        object is de-serialized, so under some circumstances, setup that happened there
+        may also need to occur here.
+
+        Args:
+            state: the object state to restore. This is effectively the output
+                of :meth:`pudl.output.pudltabl.PudlTabl.__getstate__`.
+        """
+        try:
+            pudl_engine = sa.create_engine(state["pudl_engine"])
+            # make sure that the URL for the engine from ``state`` is usable now, if
```

Comment:
I think serializing sqlalchemy to a DB connection URL is pretty clever!

Regarding the fallback behavior: typically, I think it's important to have your serialization/deserialization be inverse operations of each other, and this breaks that guarantee. However, I can see how this would be very ergonomic in some cases, like passing pickled objects back and forth across different users. I think we can keep the fallback behavior, but we should alert the user with some sort of log, so that the user can see "Oh, I'm not getting *exactly* the same object back."

Reply:
That's a great point, I'll add a logger message to that effect.

Reply:
Another thought, should something like this happen with `Datastore.local_cache_path`?